phylum/connector.lisp

```diff
@@ -0,0 +1,215 @@
+                (call-handler-helper-recurse-i (+ i 1))))]
+
+       [call-handler (resp)
+         (call-handler-helper-recurse-i 0)])
+ 
+    (lambda (op &rest args)
+        (cond ((equal? op 'register-handler) (apply register-handler args))
+              ((equal? op 'register-request-callback) (apply
+                                                        register-request-callback
+                                                        args))
+              ((equal? op 'call-handler) (apply call-handler args))
+              (:else (error 'unknown-operation op)))))))
+
+(set 'connector-handlers (singleton mk-connector-handler))
+
+;; internal handler called by the connector hub
+(defendpoint "$ch_callback" (resp)
+  (cc:infof resp "connectorhub callback")
+  (connector-handlers 'call-handler resp)
+  (route-success (sorted-map "status" "OK")))
+
+;; add-connector-event inspects an event and sets up the data structures to 
+;; register callbacks.
+(set 'add-connector-event
+  ((lambda ()
+    (let ([state (sorted-map "ctr" 0)])
+      (lambda (event &optional handler-name)
+        (let* ([ctr (get state "ctr")]
+               [event-num (get state "ctr")]
+               [event-body (get event "req")]
+               [event-req-id (or
+                               (get event-body "request_id")
+                               (mk-uuid))]
+               [event-msp (get event "msp")] ; msp ID for the connector
+               [event-key (get event "key")] ; if key omitted then use req id
+               [event-pdc (get event "pdc")] ; pdc storing key with req
+               [event-oid (or ; ID of the object
+                                   (get event "oid")
+                                   (error 'missing-oid
+                                          "missing object id"))]
+               [event-header (sorted-map "rid" event-req-id)]
+               [ctx (sorted-map "oid" event-oid
+                                "msp" event-msp
+                                "key" event-key
+                                "pdc" event-pdc)]
+               [event-ref-str
+                 (thread-first
+                   event-header
+                   (denil-map)
+                   (json:dump-bytes)
+                   (to-string))]
+               [event-ref-key (format-string "$connector_events:{}" ctr)]
+               [event-body-bytes (json:dump-bytes event-body)])
+          (when (>= ctr 10)
+            (error 'too-many-events "too many events"))
+          (when handler-name
+            (connector-handlers 
+              'register-request-callback
+              event-req-id
+              handler-name
+              ctx))
+          (cc:set-tx-metadata event-ref-key event-ref-str)
+          (if event-pdc
+            (cc:storage-put-private event-pdc
+                                    event-key
+                                    event-body-bytes)
+            (cc:storage-put event-header-key event-body-bytes))
```

Comment:
`event-header-key` appears to be undefined

Reply:
will fix